### PR TITLE
[webview_flutter] Add support for detecting process termination on Android - #130297

### PR DESCRIPTION
**Pull request by Sinishisimus**
Added a new platform interface method to PlatformNavigationDelegate in webview_flutter_platform_interface setOnWebViewRenderProcessTerminated(ProcessTerminationDetails) which handles web view render process termination on Android, for details see: https://developer.android.com/reference/android/webkit/WebViewClient#onRenderProcessGone(android.webkit.WebView,%20android.webkit.RenderProcessGoneDetail)

This PR fixes the following issue: https://github.com/flutter/flutter/issues/130297

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

**Comment by stuartmorgan** [comment]
Thanks for the contribution! You haven’t checked off any of the items in the PR checklist above, so I’m assuming this is a work in progress and am marking it as a Draft. Please review the checklist, updating the PR as appropriate, and when that’s complete please feel free to mark the PR as ready for review.

**Comment by Sinishisimus** [comment]
Hey @stuartmorgan I went through all the items in the checklist and implemented the changes needed. It could be that I missed something or that someting is off. This is where I would need yopur feedback. :)

**Comment by stuartmorgan** [comment]
It does not appear that you followed https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changing-federated-plugins, which is causing almost everything in CI to fail.

**Comment by bparrishMines** [comment]
Also, work on this will be blocked by https://github.com/flutter/flutter/issues/134777. The code that wraps the native function will need to be reworked after `webview_flutter_android` switches to the generator.

**Comment by stuartmorgan** [comment]
> I can not figure out why.

Have you looked at the `stdout` for each of the failing steps? Each of them has an explanation of what the problem is.

**Comment by Sinishisimus** [comment]
> @Sinishisimus Is this ready for review? It looks like there are still some minor `repo_checks` failures; is that the only reason it's still in Draft mode? (The version/changelog check isn't a big deal for the combo PR, but format and license should pass.)

Hey @stuartmorgan, yes, you are correct, the only reason why the PR is still in draft state is because of these minor `repo_checks` failures, which I haven't had time to address yet (but planing to address this this week hopefully). Can you maybe give some guidance how to quickly address the remainng issues you mentioned (but format and license should pass)?
I see olny Linux repo checks not passing https://github.com/flutter/packages/pull/6508/checks?check_run_id=24301966434

**Comment by bparrishMines** [comment]
@Sinishisimus To fix formatting, you can run the tool from the root of the repository:

```bash
( cd script/tool/ ; dart pub get )
dart run script/tool/bin/flutter_plugin_tools.dart format
```

For the license and other failures, you can click on `details` and then `View more details on flutter-dashboard` of the `Linux repo_checks`. These should give you an explanation on how to fix them. You can ignore the failures for `CHANGELOG and version validation` and `federated safety check` for now since they should be fail until this PR is split.

File: packages/webview_flutter/webview_flutter/CHANGELOG.md [hunk 1]
```diff
@@ -1,3 +1,10 @@
+## 4.8.1
+
+* Adds support for detecting web view render process termination on Android via navigation delegate. See
+  `NavigationDelegate.onWebViewRenderProcessTerminated`.
+  Calling application should return `true` if termination was handled by the app, otherwise `false`.
+  For more details see: https://developer.android.com/reference/android/webkit/WebViewClient#onRenderProcessGone(android.webkit.WebView,%20android.webkit.RenderProcessGoneDetail)
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/CHANGELOG.md:6]: We do not generally want to put Android-specific docs in the app-facing package. Our cross-platform abstraction needs to be understandable without going to platform-level docs.

File: packages/webview_flutter/webview_flutter/CHANGELOG.md [hunk 2]
```diff
@@ -1,3 +1,10 @@
+## 4.8.1
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/CHANGELOG.md:1]: New features require a minor version bump, per semver.

File: packages/webview_flutter/webview_flutter/lib/src/navigation_delegate.dart [hunk 3]
```diff
@@ -183,4 +196,9 @@ class NavigationDelegate {
 
   /// Invoked when a resource loading error occurred.
   final WebResourceErrorCallback? onWebResourceError;
+
+  /// Invoked when a the given WebView's render process has exited (Android only).
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter/lib/src/navigation_delegate.dart:200]: Same.

File: packages/webview_flutter/webview_flutter_android/CHANGELOG.md [hunk 4]
```diff
@@ -1,3 +1,8 @@
+## 3.16.3
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/CHANGELOG.md:1]: Minor version, per previous comment.

File: packages/webview_flutter/webview_flutter_android/CHANGELOG.md [hunk 5]
```diff
@@ -1,3 +1,8 @@
...
+
+* Adds onWebViewRenderProcessTerminated WebViewClient callback on Android to support
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/CHANGELOG.md:3]: Changelog entries in an Android-specific package don't need to say "on Android".

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart [hunk 6]
```diff
@@ -932,6 +934,25 @@ class WebViewClient extends JavaObject {
     WebResourceError error,
   )? onReceivedRequestError;
 
+  /// Notify host application that the given WebView's render process has exited (Android only).
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:937]: This package is Android-only.

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart [hunk 7]
```diff
@@ -932,6 +934,25 @@ class WebViewClient extends JavaObject {
...
+  /// See: https://developer.android.com/reference/android/webkit/WebViewClient#onRenderProcessGone(android.webkit.WebView,%20android.webkit.RenderProcessGoneDetail)
+  ///
+  /// Multiple WebView instances may be associated with a single render process;
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:940]: Please don't copy and paste large amounts of text from the Android docs.

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart [hunk 8]
```diff
@@ -1512,6 +1553,23 @@ class WebResourceError {
   final String description;
 }
 
+/// This class provides more specific information about why the webview render process exited (Android only).
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:1556]: As above, no need to say "Android only".

File: packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart [hunk 9]
```diff
@@ -1512,6 +1553,23 @@ class WebResourceError {
...
+/// The application may use this to decide how to handle the situation.
+/// See [WebViewClient.onWebViewRenderProcessTerminated].
+class ProcessTerminationDetails {
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_android/lib/src/android_webview.dart:1559]: Why is the same class being declared in two different packages?

File: packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md [hunk 10]
```diff
@@ -1,5 +1,6 @@
-## NEXT
+## 2.10.1
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_platform_interface/CHANGELOG.md:1]: Minor version.

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_navigation_delegate.dart [hunk 11]
```diff
@@ -128,6 +132,16 @@ abstract class PlatformNavigationDelegate extends PlatformInterface {
         'setOnWebResourceError is not implemented on the current platform.');
   }
 
+  /// Notify host application that the given WebView's render process has exited.
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/platform_navigation_delegate.dart:135]: Please use the same "Invoked when ..." structure as all the other comments here.

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/process_termination_details.dart [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+@immutable
+class ProcessTerminationDetails {
+  /// Creates a [ProcessTerminationDetails].
+  const ProcessTerminationDetails({
+    required this.didCrash,
+    required this.rendererPriorityAtExit,
+  });
+
+  /// Indicates whether the render process was observed to crash, or whether it was killed by the system.
+  final bool didCrash;
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/process_termination_details.dart:18]: What would a platform implementation that didn't have access to this information pass? It seems like this needs to be an enum, where one of the options is "unknown".

File: packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/process_termination_details.dart [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+
+  /// Returns the renderer priority that was set at the time that the renderer exited.
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_platform_interface/lib/src/types/process_termination_details.dart:20]: This is extremely Android-specific, so should not be in this package. If it's actually needed it should be in an Android-specific subclass, but it's not clear to me that there's a plausible use case for this data.

File: packages/webview_flutter/webview_flutter_web/example/pubspec.yaml [hunk 14]
```diff
@@ -28,3 +28,8 @@ dev_dependencies:
 
 flutter:
   uses-material-design: true
+
+# FOR TESTING AND INITIAL REVIEW ONLY. DO NOT MERGE.
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_web/example/pubspec.yaml:32]: These changes can be reverted in packages that have no other changes.

File: packages/webview_flutter/webview_flutter_wkwebview/CHANGELOG.md [hunk 15]
```diff
@@ -1,3 +1,8 @@
+## 3.13.2
+
+* Adds `decidePolicyForNavigationResponse` to internal WKNavigationDelegate to support the
```
Comment (stuartmorgan) [inline packages/webview_flutter/webview_flutter_wkwebview/CHANGELOG.md:3]: This is copy/pasted from a previous changelog entry, and is unrelated to the PR.

**Review by stuartmorgan — CHANGES_REQUESTED** [review]
Where is the iOS implementation? Please see the original description of https://github.com/flutter/flutter/issues/130297, as well as https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#platform-support

**Comment by yaakovschectman** [comment]
From Android triage: We will close this for now to get it off of our review queue while there are open comments and merge conflicts. Please reopen or open a new PR if you get back to this.

**Pull request closed without merging**